CMakeLists.txt

```
@@ -615,7 +615,33 @@ set(MASTER_HEADERS
   ${PROJECT_SOURCE_DIR}/mfem.hpp
   ${PROJECT_SOURCE_DIR}/mfem-performance.hpp)
 
-set(_lib_path "${CMAKE_INSTALL_PREFIX}/lib")
+# Installation options
+set(MFEM_USE_GNUINSTALLDIRS OFF
+  CACHE BOOL "Enable to use the GNUInstallDirs CMake module.")
+set(install_include_dir_default include)
+set(install_bin_dir_default bin)
+set(install_lib_dir_default lib)
+set(install_share_dir_default share)
+if (MFEM_USE_GNUINSTALLDIRS)
+  include(GNUInstallDirs)
+  set(install_include_dir_default "${CMAKE_INSTALL_INCLUDEDIR}")
+  set(install_bin_dir_default "${CMAKE_INSTALL_BINDIR}")
+  set(install_lib_dir_default "${CMAKE_INSTALL_LIBDIR}")
+  set(install_share_dir_default "${CMAKE_INSTALL_DATAROOTDIR}")
+endif()
+mfem_cache_path(INSTALL_INCLUDE_DIR "${install_include_dir_default}"
+  "Relative or absolute path for installing header files.")
+mfem_cache_path(INSTALL_BIN_DIR "${install_bin_dir_default}"
+  "Relative or absolute path for installing the binaries.")
+mfem_cache_path(INSTALL_LIB_DIR ${install_lib_dir_default}
+  "Relative or absolute path for installing the library.")
+mfem_cache_path(INSTALL_SHARE_DIR "${install_share_dir_default}"
+  "Relative or absolute path for installing shared data.")
+# other options: "share/mfem/cmake", "lib/mfem/cmake"
+mfem_cache_path(INSTALL_CMAKE_DIR ${INSTALL_LIB_DIR}/cmake/mfem
+  "Relative or absolute path for installing cmake config files.")
+
+mfem_path_to_fullpath("${INSTALL_LIB_DIR}" "${CMAKE_INSTALL_PREFIX}" _lib_path)
 set(CMAKE_INSTALL_RPATH_USE_LINK_PATH ON CACHE BOOL "")
 set(CMAKE_INSTALL_RPATH "${_lib_path}" CACHE PATH "")
 set(CMAKE_INSTALL_NAME_DIR "${_lib_path}" CACHE PATH "")
```

```suggestion
set(CMAKE_INSTALL_RPATH "${_lib_path}" CACHE INTERNAL "" FORCE)
set(CMAKE_INSTALL_NAME_DIR "${_lib_path}" CACHE INTERNAL "" FORCE)
```

Is there anytime these should be different from `${CMAKE_INSTALL_PREFIX}/${INSTALL_LIB_DIR}`? Without FORCE (alternative: remove the `CACHE`) changing `CMAKE_INSTALL_PREFIX` won't update these without the user deleting these two entries from the cache.

---

Similar to this, `MFEM_SOURCE_DIR` and `MFEM_INSTALL_DIR` should either be marked `CACHE INTERNAL FORCE` or just not put into `CACHE` at all.

---

> Is there anytime these should be different from ${CMAKE_INSTALL_PREFIX}/${INSTALL_LIB_DIR}?

Yes, e.g. `INSTALL_LIB_DIR` can hold an absolute path, completely unrelated to `CMAKE_INSTALL_PREFIX`.

---

> Regarding MFEM_SOURCE_DIR and MFEM_INSTALL_DIR: I agree, I don't think there's a reason for them to not be made INTERNAL or just not stored in the cache.

You want me to do this here in this PR? I'm a little sad that I did not just use the existing variables in the beginning to accomplish the goal of this PR, which is installing the common headers and library. This PR has really exploded in terms of discussion beyond the scope I intended

---

@lindsayad, I apologize, you are right -- the scope of the changes that we (the reviewers) started to ask for grew far beyond the original scope.

In order to minimize any additional effort on your side, I can take care of adding any final changes, if that is okay with you? Since the PR branch is in a fork, you may need to give me permission to write to your brach -- there should be a checkbox for that in the right panel of the PR page.

---

that works great for me, thanks @v-dobrev. Yes, as a maintainer myself, I always grant access to maintainers 😄

---

If we want these to be optionally user configurable, then I think we should use this:

```suggestion
if(NOT DEFINED CMAKE_INSTALL_RPATH)
  set(CMAKE_INSTALL_RPATH "${_lib_path}")
endif()
if(NOT DEFINED CMAKE_INSTALL_NAME_DIR)
  set(CMAKE_INSTALL_NAME_DIR "${_lib_path}")
endif()
```

This way, if the user doesn't manually specify one or both of these CMake will automatically track them to `${_lib_path}`, and if the user does set them CMake will use the user-defined values.

One time I can think it makes sense to not use `${_lib_path}` is to allow the installed binary (and shared libmfem) to be relocated after install (say as part of an installer or dependency of another package), in which case maybe it makes sense to default these to `$ORIGIN/../${INSTALL_LIB_DIR}`?

---

> This way, if the user doesn't manually specify one or both of these CMake will automatically track them to `${_lib_path}`, and if the user does set them CMake will use the user-defined values.

Value set on the command line with `-DCMAKE_INSTALL_RPATH=...` are cached and will not be overwritten by `set(CMAKE_INSTALL_RPATH "${_lib_path}" CACHE PATH "")` -- this way we do not need to explicitly check if `CMAKE_INSTALL_RPATH` is defined. This will also work if one adds their own setting to `config/user.cmake` with `set(CMAKE_INSTALL_RPATH "${my_rpath}" CACHE PATH "")`. Checking if `CMAKE_INSTALL_RPATH` is defined makes sense only if we want to support the case where `CMAKE_INSTALL_RPATH` is defined as a non-cache variable. Is that the reason you want to have the explicit `if (NOT DEFINED ...)` check?

> One time I can think it makes sense to not use `${_lib_path}` is to allow the installed binary (and shared libmfem) to be relocated after install (say as part of an installer or dependency of another package), in which case maybe it makes sense to default these to `$ORIGIN/../${INSTALL_LIB_DIR}`?

Yes, I think this may be one use case. Another case could be when a user wants to add more paths to the rpath for some reason.

---

@v-dobrev, I think @helloworld922 meant that if you run CMake once with a given `CMAKE_INSTALL_PATH` and then a subsequent time (without deleting the contents of the directory you're running CMake from) with a different `CMAKE_INSTALL_PATH`, then these two variables will not reflect that change and will be stuck with the path computed during the first CMake pass.

---

Yes, the situation I was envisioning is if you call cmake initially then later call `cmake -DCMAKE_INSTALL_PREFIX=...` (or use `ccmake`/`cmake-gui`to modify these) it won't reflect these changes in `CMAKE_INSTALL_RPATH` and `CMAKE_INSTALL_NAMED_DIR` even if the user never intended to set them initially.

---

Okay, I see. I'm not sure we fully support re-configuration with CMake as you describe it. I know it is something CMake wants to support but it typically creates complications such as what you describe.

I never trust re-configuration (my impression is that most projects, including mfem, have issues with that), so I always delete everything in the CMake build directory before running `cmake` again.

That said, what you suggest seems a step in the right direction to support re-configuration, so let's add it -- I'll commit your suggestion.